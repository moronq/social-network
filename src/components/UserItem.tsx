import React, { FC } from 'react'
import { Link } from 'react-router-dom'
import { usersAPI } from '../api/api'
import { useAppDispatch } from '../hooks/redux'
import noAvatarSmall from './../assets/img/no-avatar-small.png'

type UserItemType = {
  photo: string | null
  name: string
  status: string | null
  followed: boolean
  id: number
}

const UserItem: FC<UserItemType> = ({ photo, name, status, followed, id }) => {
  const onFollowClick = () => {
    usersAPI.follow(id)
  }

  return (
    <li>
      <Link to={`/profile/${id}`}>
        <img src={photo || noAvatarSmall} alt="user-avatar" />
      </Link>
      <p>{name}</p>
      <p>{status || 'Не придумал статус'}</p>
      <button
        onClick={onFollowClick}
        className="bg-slate-300 px-2 hover:bg-slate-100"
      >
        {followed ? 'unfollow' : 'follow'}
      </button>
    </li>
  )
}

export default UserItem
