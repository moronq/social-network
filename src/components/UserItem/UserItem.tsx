import React, { FC, useState } from 'react'
import { Link } from 'react-router-dom'
import { usersAPI } from '../../api/api'
import { useAppDispatch, useAppSelector } from '../../hooks/redux'
import { follow, unfollow } from '../../store/slices/usersSlice/usersAction'
import noAvatarSmall from '../../assets/img/no-avatar-small.png'

type UserItemType = {
  photo: string | null
  name: string
  status: string | null
  followed: boolean
  id: number
}

const UserItem: FC<UserItemType> = ({ photo, name, status, followed, id }) => {
  const dispatch = useAppDispatch()
  const [waiting, setWaiting] = useState(false)
  const onFollowClick = () => {
    if (!followed) {
      setWaiting(true)
      dispatch(follow(id)).then(() => setWaiting(false))
    } else {
      setWaiting(true)
      dispatch(unfollow(id)).then(() => setWaiting(false))
    }
  }

  return (
    <li>
      <Link to={`/profile/${id}`}>
        <img src={photo || noAvatarSmall} alt="user-avatar" />
      </Link>
      <p>{name}</p>
      <p>{status || 'Не придумал статус'}</p>
      <button
        disabled={waiting}
        onClick={onFollowClick}
        className="bg-slate-300 px-2 hover:bg-slate-100 disabled:text-white"
      >
        {followed ? 'unfollow' : 'follow'}
      </button>
    </li>
  )
}

export default UserItem
