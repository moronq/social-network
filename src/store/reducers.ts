import auth from './slices/authSlice/authSlice'
import profile from './slices/profileSlice/profileSlice'

export default {
  auth,
  profile,
}
