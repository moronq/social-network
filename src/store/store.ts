import { combineReducers, configureStore } from '@reduxjs/toolkit'
import reducers from './reducers'

const rootReducer = combineReducers(reducers)

export const store = configureStore({
  reducer: rootReducer,
})
