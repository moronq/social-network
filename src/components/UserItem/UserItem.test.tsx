import { render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom'
import { store } from '../../store/store'
import UserItem from './UserItem'

const mockUser = {
  followed: false,
  id: 1,
  name: 'John',
  photo: null,
  status: 'user status',
}

describe('UserItem test group', () => {
  test('UserItem renders without errors', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <UserItem
            followed={mockUser.followed}
            id={mockUser.id}
            name={mockUser.name}
            photo={mockUser.photo}
            status={mockUser.status}
          />
        </MemoryRouter>
      </Provider>
    )
    expect(screen.getByText('John')).toBeInTheDocument
    expect(screen.getByText('user status')).toBeInTheDocument
  })

  test('snapshot test', () => {
    const userItemComponent = render(
      <Provider store={store}>
        <MemoryRouter>
          <UserItem
            followed={mockUser.followed}
            id={mockUser.id}
            name={mockUser.name}
            photo={mockUser.photo}
            status={mockUser.status}
          />
        </MemoryRouter>
      </Provider>
    )
    expect(userItemComponent).toMatchSnapshot()
  })
})
