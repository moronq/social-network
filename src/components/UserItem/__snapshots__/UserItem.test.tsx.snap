// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserItem test group snapshot test 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li>
        <a
          href="/profile/1"
        >
          <img
            alt="user-avatar"
            src="no-avatar-small.png"
          />
        </a>
        <p>
          John
        </p>
        <p>
          user status
        </p>
        <button
          class="bg-slate-300 px-2 hover:bg-slate-100 disabled:text-white"
        >
          follow
        </button>
      </li>
    </div>
  </body>,
  "container": <div>
    <li>
      <a
        href="/profile/1"
      >
        <img
          alt="user-avatar"
          src="no-avatar-small.png"
        />
      </a>
      <p>
        John
      </p>
      <p>
        user status
      </p>
      <button
        class="bg-slate-300 px-2 hover:bg-slate-100 disabled:text-white"
      >
        follow
      </button>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
